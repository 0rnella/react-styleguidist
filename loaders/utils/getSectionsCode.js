'use strict';

// This two functions should be in the same file because of cyclic imports

const path = require('path');
const toCode = require('./toCode');
const requireIt = require('./requireIt');
const getComponents = require('./getComponents');
const getComponentsCode = require('./getComponentsCode');

/**
 * Return JS code as a string for one level of sections.
 *
 * @param {Array} sections
 * @param {object} config
 * @returns {string|null}
 */
function getSectionsCode(sections, config) {
	if (!sections) {
		return null;
	}

	return toCode(sections.map(section => processSection(section, config)));
}

/**
 * Return JS code as a string for a given section with all components and subsections.
 * @param {object} section
 * @param {object} config
 * @returns {string}
 */
function processSection(section, config) {
	return toCode({
		name: JSON.stringify(section.name),
		content: section.content
			? requireIt('!!examples-loader!' + path.resolve(config.configDir, section.content))
			: null,
		components: getComponentsCode(getComponents(section.components, config), config),
		sections: getSectionsCode(section.sections, config),
	});
}

module.exports = getSectionsCode;
module.exports.processSection = processSection;
