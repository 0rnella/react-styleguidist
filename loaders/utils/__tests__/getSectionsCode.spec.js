import vm from 'vm';
import path from 'path';
import getSectionsCode, { processSection } from '../getSectionsCode';

const sections = [
	{
		name: 'Readme',
		content: 'Readme.md',
	},
	{
		name: 'Components',
		components: 'components/**/[A-Z]*.js',
	},
];
const config = {
	configDir: __dirname,
	getExampleFilename: componentpath => path.join(path.dirname(componentpath), 'Readme.md'),
	getComponentPathLine: componentpath => componentpath,
};

const removeDirname = string => string.replace(new RegExp(__dirname, 'g'), '.');

it('processSection() should return valid JS for section with content', () => {
	const result = processSection(sections[0], config);

	expect(new vm.Script(`a=${result}`)).not.toThrowError(SyntaxError);
	expect(removeDirname(result)).toMatchSnapshot();
});

it('processSection() should return valid JS for section with components', () => {
	const result = processSection(sections[1], config);

	expect(new vm.Script(`a=${result}`)).not.toThrowError(SyntaxError);
	expect(removeDirname(result)).toMatchSnapshot();
});

it('getSectionsCode() should return null if there is no sections', () => {
	const result = getSectionsCode();

	expect(result).toBeNull();
});

it('getSectionsCode() should return valid JS', () => {
	const result = getSectionsCode(sections, config);

	expect(new vm.Script(result)).not.toThrowError(SyntaxError);
	expect(removeDirname(result)).toMatchSnapshot();
});
