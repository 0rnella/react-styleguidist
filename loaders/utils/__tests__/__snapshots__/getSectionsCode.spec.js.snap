exports[`test getSectionsCode() should return valid JS 1`] = `
"[{\'name\': \"Readme\",
\'content\': require(\"!!../loaders/examples-loader!./Readme.md\"),
\'components\': null,
\'sections\': null},
{\'name\': \"Components\",
\'content\': null,
\'components\': null,
\'sections\': null}]"
`;

exports[`test processSection() should return valid JS for section with components 1`] = `
"{\'name\': \"Components\",
\'content\': null,
\'components\': null,
\'sections\': null}"
`;

exports[`test processSection() should return valid JS for section with content 1`] = `
"{\'name\': \"Readme\",
\'content\': require(\"!!../loaders/examples-loader!./Readme.md\"),
\'components\': null,
\'sections\': null}"
`;
